### ISSUE TITLE
"reportIncompatibleVariableOverride" when overriding an immutable field type with an immutable subtype - False positive?

### ISSUE THREAD

**Issue by mikenerone** [microsoft/pyright#6270]
I'll first mention that I'm not 100% sure this is a bug. Please tell me. :)

The issue I suspect is that Pyright treats type unions as mutable even if all of the subtypes are immutable. Please consider this example with Pydantic 2.4.2 (of course, my real-world case is significantly more complicated):
```python
from uuid import UUID
from pydantic import BaseModel

class IssueUpdate(BaseModel):
    guid: UUID
    summary: str | None = None
    description: str | None = None
    category: str | None = None

class Issue(IssueUpdate):
    summary: str
    description: str
    category: str
```
In case it's not clear, what I'm after here is two models that represent slightly different forms of the same entity. In principle, all of these fields are required, but the "update" model is used to represent a delta to be applied to an existing object. Since the existing object is fully populated, many of these fields can be treated as optional.

For the above, Pyright 1.1.333 issues this warning:
```
...../example.py
  ...../example.py:13:5 - error: "summary" overrides symbol of same name in class "IssueUpdate"
    Variable is mutable so its type is invariant
      Override type "str" is not the same as base type "str | None" (reportIncompatibleVariableOverride)
  ...../example.py:14:5 - error: "description" overrides symbol of same name in class "IssueUpdate"
    Variable is mutable so its type is invariant
      Override type "str" is not the same as base type "str | None" (reportIncompatibleVariableOverride)
  ...../example.py:15:5 - error: "category" overrides symbol of same name in class "IssueUpdate"
    Variable is mutable so its type is invariant
      Override type "str" is not the same as base type "str | None" (reportIncompatibleVariableOverride)
3 errors, 0 warnings, 0 informations
```
I would submit that because `str` and `None` are both immutable, `str | None` is also immutable, and as such, this field should actually be treated as covariant. If I'm mistaken, please enlighten me (and maybe someone could toss in a quick tip as to how to properly accomplish my goal here, preferably not simply ignoring the warning).

**VS Code extension or command-line**
Command-line pyright via pre-commit.

**Comment by erictraut** [comment]
"Immutability" isn't a property of a type. It's a property of a variable. It means "once the initial value of the variable is assigned, it cannot be subsequently changed".

The following sample demonstrates why the code above is not type safe.

```python
# pyright: strict

from pydantic import BaseModel

class IssueUpdate(BaseModel):
    summary: str | None = None

class Issue(IssueUpdate):
    summary: str

def clear_summary(x: IssueUpdate):
    x.summary = None

i = Issue(summary="Must be a str")
assert isinstance(i.summary, str)

clear_summary(i)
assert isinstance(i.summary, str)  # No longer true!
```

In this example, the instance variable `IssueUpdate.summary` is mutable — i.e. it can be modified. That means someone can come along and set it to a new value that is `None`.

**Comment by mikenerone** [comment]
Ahh, I see it now. Ok, thanks.

**Comment by mikenerone** [comment]
I think I might have simply messed up my example. What if we reverse the inheritance?
```python
from pydantic import BaseModel

class Issue(BaseModel):
    summary: str

class IssueUpdate(Issue):
    summary: str | None = None
```
pyright produces the same warning with this structure, but I think this actually fixes the type safety. A mutating function respecting the annotations, acting on either type, can't violate:
```python
def change_an_issue(i: Issue):
    i.summary = "a string"  # No problem, string is fine for both the base class and the subclass.
    # or
    i.summary = None  # This would produce a warning just as we'd hope (the fact that it would actually be ok for the subclass is immaterial).
    
def change_an_issue_update(i: IssueUpdate):
    i.summary = "a string"  # No problem, string is fine for IssueUpdate
    # or
    i.summary = None  # Also no problem, None is fine for IssueUpdate
```

**Comment by erictraut** [comment]
That's not type safe either.

```python
# pyright: strict

from pydantic import BaseModel

class Issue(BaseModel):
    summary: str

    def print_summary(self):
        print(self.summary.capitalize())

class IssueUpdate(Issue):
    summary: str | None = None

i = IssueUpdate()
i.print_summary()  # Crash!
```

**Comment by mikenerone** [comment]
Oof. :)
